Admin Menu: Make Home & Stats visible in Locked Mode

obenland opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

For sites on 100-Year plans that have Locked Mode enabled, the Home and Stats menu item are currently removed because we disable the `edit_posts` capability.

On Simple sites, all available roles have `edit_posts` and `read` capabilities, so there is no difference in functionality in using `read` over `edit_posts` for the My Home and Stats menu items.

On WoA sites, the My Home menu item already uses the `read` capability. For stats we can use Jetpack's `view_stats` capability and filter the default options for the Stats menu to maintain access for all roles.

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Update Stats menu item capability on WoA sites to `view_stats` to re-use Jetpack capability for that purpose.
* Add a default option callback to maintain access to Stats for all roles.
* Update My Home menu capability on Simple Sites to `read` to bring in line with Core.
* Update Stats menu item capability on Simple Sites to `read` so it doesn't get hidden when we remove the `edit_posts` capability.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Apply the diff to your sandbox and sandbox public-api.
* Apply the diff to a WoA site using Jetpack Beta.
* Test admin menu with simple and WoA sites with admin and contributor roles.
* Make sure My Home and Stats are always accessible.
* Upgrade a test site to the 100-year plan.
* In Settings General, enable Locked Mode.
* Make sure My Home and Stats are always accessible.

obenland committed:
Admin Menu: Make Home & Stats visible in Locked Mode

obenland committed:
There's no view_stats on Simple

read is a similar-level cap as edit_posts on Simple

mmtr reviewed (dismissed):
> Changing Home from edit_posts to read will enable the Calypso My Home screeen for all users that currently have a Subscriber role. I'm not sure how big of a problem that is, but we should be at least aware of it.

I was concerned a little bit about this at first, but seems this is harmless since Subscribers don't have access to sites they are subscribed to from the "My Site(s)" menu.

> Changing Stats from edit_posts to view_stats will remove the Stats menu item for Editors, Authors, and Contributors

To clarify, this is only for Atomic sites, right? According to the code (and my tests) users of Simple sites will still have access to Stats as long as they have the `read` capability.

Which I think is the correct behavior, because Simple Sites don't have ([the ability to choose which roles can see stats](https://wordpress.com/support/stats/#authors)) so we need to ensure that Editors, Authors, and Contributors of Simple sites can continue accessing Stats.

obenland commented:
Yes, apologies, I've not had a chance to update the issue description. 

There are no subscriber-level roles on Simple, so every one has `edit_posts` AND `read`, so they can be used "interchangeably".

> Simple Sites don't have ([the ability to choose which roles can see stats](https://wordpress.com/support/stats/#authors))

It took me a while to figure that out on Friday. So yeah, with `read` there should be no changes on Simple. For WoA I think I'd like to at least add a filter to the default option value for Stats to not limit it to Admins only. Not sure if that's possible, but that'll be my next step on this.

obenland committed:
Maintain access to Stats menu item for all roles

Can be changed by site owners later, but this way access on WoA sites will start out the way it is on Simple.

obenland committed:
Merge branch 'trunk' into update/menu-caps

mmtr commented on projects/plugins/jetpack/modules/masterbar/admin-menu/class-atomic-admin-menu.php:
```diff
@@ -444,4 +445,18 @@ public function wp_ajax_jitm_dismiss() {
+	/**
+	 * Extends the default value for which roles can access the Stats menu item.
+	 *
+	 * @see Automattic\Jetpack\Stats\Options::get_default_value()
```
`get_default_value` does not exist in `Automattic\Jetpack\Stats\Options`. Maybe you meant `Automattic\Jetpack\Stats\Options::get_options()`?

See https://github.com/Automattic/jetpack/blob/f58c02baffc449b0bebd70421c8234e4740cea01/projects/packages/stats/src/class-options.php#L42

obenland replied:
It's `get_defaults()`! Gah, Copilot!

https://github.com/Automattic/jetpack/blob/f58c02baffc449b0bebd70421c8234e4740cea01/projects/packages/stats/src/class-options.php#L156

mmtr reviewed (commented):
Nice, this works as expected.

> Add a default option callback to maintain access to Stats for all roles.

This has the side effect of showing the Jetpack Stats section in Jetpack > Dashboard (`/wp-admin/admin.php?page=jetpack#/dashboard`) to contributors, authors, and editors, which was previously hidden to them:

Before | After
--- | ---
<img width="1385" alt="Screenshot 2023-09-26 at 13 52 17" src="https://github.com/Automattic/jetpack/assets/1233880/37d032dc-39a1-4f34-9d44-a41c0d02fc16"> | <img width="1388" alt="Screenshot 2023-09-26 at 13 53 02" src="https://github.com/Automattic/jetpack/assets/1233880/648af003-8dcd-46a4-a128-e93d910eb3ae">

I guess it sort of makes sense, since these users have access to the stats data anyway via  Calypso (`/stats/day/:site`), but I wonder if the Nav Unification logic is the best place to change the default roles since it's not immediately obvious that this also affects the Jetpack Dashboard.

obenland commented:
> This has the side effect of showing the Jetpack Stats section in Jetpack > Dashboard

Nice catch! Agreed, I think of it more as fixing a bug than an unintended side effect. The bar graph in the sidebar now works, too!

obenland committed:
Merge branch 'trunk' into update/menu-caps

obenland committed:
Fix typo

Props @mmtr.

obenland commented:
> I wonder if the Nav Unification logic is the best place to change the default roles since it's not immediately obvious that this also affects the Jetpack Dashboard.

I kept in the class because that's where we're using the capability. Would wpcomsh be a better place?

mmtr commented:
> Would wpcomsh be a better place?

I think so, yeah.

obenland committed:
Remove cap callback.

This will live in wpcomsh in the future.

obenland committed:
Merge branch 'trunk' into update/menu-caps

mmtr reviewed (approved):
Works great when paired with https://github.com/Automattic/wpcomsh/pull/1509

obenland committed:
Merge branch 'trunk' into update/menu-caps

obenland commented:
@anomiex Is it okay to merge with these tests failing? Any idea how this diff would effect those?

obenland commented:
Hm, looks like they're failing on trunk, too

anomiex commented:
Yes, you should be ok. Details on the PHP test failure are in pdWQjU-zp-p2. The other one appears to be an issue with Actions rolling out an update from node 18.17.1 to 18.18.0 which apparently breaks things, I'm working on a workaround for that now.
